File: src.csharp/AlphaTab/Core/EcmaScript/Set.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace AlphaTab.Core.EcmaScript;

public abstract class Set
{
}

public class Set<T> : Set, IEnumerable<T>, ICollection
{
    private readonly HashSet<T> _data;

    public Set()
    {
        _data = new HashSet<T>();
    }

    public double Size => _data.Count;

    public Set(IEnumerable<T>? values)
    {
        _data = values == null ? new HashSet<T>() : new HashSet<T>(values);
    }

    public void Clear()
    {
        _data.Clear();
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Add(T item)
    {
        _data.Add(item);
    }


    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Has(T item)
    {
        return _data.Contains(item);
    }

    public void ForEach(Action<T> action)
    {
        foreach (var i in _data)
        {
            action(i);
        }
    }

    public IEnumerator<T> GetEnumerator()
    {
        return _data.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public void Delete(T item)
    {
        _data.Remove(item);
    }

    int ICollection.Count => _data.Count;
    bool ICollection.IsSynchronized => false;
    object ICollection.SyncRoot => _data;
    void ICollection.CopyTo(System.Array array, int index)
    {
        _data.CopyTo((T[])array, index);
    }
}

File: src.kotlin/alphaTab/alphaTab/src/commonMain/kotlin/alphaTab/core/ecmaScript/Set.kt
package alphaTab.core.ecmaScript

internal class Set<T> : Iterable<T> {
    private val _set: HashSet<T>

    public constructor() {
        _set = HashSet()
    }

    public constructor(values: Iterable<T>?) {
        _set = values?.toHashSet() ?: HashSet()
    }

    public fun add(item: T) {
        _set.add(item)
    }

    public fun has(item: T): Boolean {
        return _set.contains(item)
    }

    public fun delete(item: T) {
        _set.remove(item)
    }

    public fun forEach(action: (item: T) -> Unit) {
        for (i in _set) {
            action(i)
        }
    }

    override fun iterator(): Iterator<T> {
        return _set.iterator()
    }
}
